### PR TITLE
A functional `rustls-cert-gen` with basic parameters.

### LINKED ISSUE #175: Command line tool
I've seen some talk about the possibility of bringing a command-line-tool to rcgen. [minica](https://github.com/jsha/minica) is an example of this kind of tool for Golang. I recently wrote [myca](https://github.com/tbro/myca) to address my own needs. It uses **rcgen** library. My hope with this issue is to spark a discussion about bringing a **CLI** to **rcgen**. The first question might be to ask if  **rcgen** wants such a thing? If it does, then we can discuss what it should look like. I think it would be relatively easy to provide a tool to create throw away keys and certificates for use in development and testing. And that is what I have tried to solve for myself. Once that bridge has been crossed, it shouldn't be much more effort to provide a few options and switches for compliance in more restricted contexts. I might be oversimplifying and I'm sure there are other opinions. Discussion is welcome.

### PR DESCRIPTION
**Pull request by tbro**
This is basically #185 minus #188 and #189. The structure also differs as sub modules have been inlined in `main.rs` and `cert.rs`. `anyhow` has also been added as a dependency to replace the `Result` alias.

Closes #175

### CODE REVIEW

File: Cargo.toml
Comment (cpu) [inline Cargo.toml:9]: [My feedback](https://github.com/rustls/rcgen/pull/185/files#r1385579137) on the other PR about the `rsa` dependency was marked as resolved, but the dependency is still present. 

Do you disagree with my suggestion we consider RSA key generation out-of-scope for this tool since rcgen the library can't support it, or did something get lost in the shuffle?

File: rustls-cert-gen/README.md
Comment (cpu) [inline rustls-cert-gen/README.md:18]:
```suggestion
  * `root-ca.pem` (ca's self-signed X.509 certificate)
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:32]:
```suggestion
/// into either a [Ca] or an [EndEntity].
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:51]: Nit: I think this is duplication with the fn signature.
```suggestion
	/// Set signature algorithm (instead of default).
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:105]:
```suggestion
		Ok(Ca { cert: Certificate::from_params(self.params)? } )
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:121]:
```suggestion
		Ok(PemCertifiedKey {
			cert_pem: self.cert.serialize_pem()?,
			private_key_pem: self.cert.serialize_private_key_pem(),
		})
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:141]:
```suggestion
		Ok(PemCertifiedKey {
			cert_pem: self.cert.serialize_pem_with_signer(signer)?,
			private_key_pem: self.cert.serialize_private_key_pem(),
		})
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:163]: It looks like calling this multiple times will append multiple `CommonName` DNs. I think probably this should only set the Common Name DN once? Or it should have a docstring describing the behaviour.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:169]: I think this builder fn has the same property where it's appending for multiple calls. That seems OK for SANs but I think that should be called out in a doc string.

In an ideal world we would error on duplicate SAN values, but that might make more sense in `rcgen`'s handling of the parameters.

```suggestion
		self.params.subject_alt_names.extend(sans);
		self
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:175]:
```suggestion
self.params.extended_key_usages.push(ExtendedKeyUsagePurpose::ClientAuth);
		self
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:181]:
```suggestion
	self.params.extended_key_usages.push(ExtendedKeyUsagePurpose::ServerAuth);
		self
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:186]:
```suggestion
		Ok(EndEntity { cert: Certificate::from_params(self.params)? })
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:191]: Except for tests, the `alg` field is never read. In `CertificateBuilder::signature_algorithm` the `key_pair` field is used to populate the `params.alg`.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:190]: I think this type would be better expressed as a an enum named something like `KeypairAlgorithm` with variants `Ed25519`, `EcdsaP256` and `EcdsaP384`.

I think we should avoid "stringly-typed" inputs since they're hard for consumers to use. The `ed25519`, `ecdsa_p256` and `ecdsa_p384` functions on this type could be reconstituted as `KeypairAlgorithm::to_keypair` that converts to `rcgen::KeyPair` as appropriate.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:195]: Just noting that I'm skipping over reviewing this part based on the feedback above.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:174]: This should probably only `push` if the EKU isn't already present to avoid duplicates.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:180]: same feedback here w.r.t duplicates.

File: rustls-cert-gen/src/main.rs
Comment (cpu) [inline rustls-cert-gen/src/main.rs:92]: WDYT about in-lining this into `parse_sans`? I don't think it has much value on its own since only `parse_sans` uses it (excluding tests which could use `parse_sans` instead).

**Review by cpu — COMMENTED** [review]
Thanks! Here's another round of feedback

File: Cargo.toml
Comment (tbro) [reply Cargo.toml:9]: It looks like I removed it from the `rustls-cert-gen` crate but forgot to remove it here.

File: rustls-cert-gen/src/cert.rs
Comment (tbro) [reply rustls-cert-gen/src/cert.rs]: I agree. We can have the cli arg parser parse directly to an enum so that `to_keypair` can take a variant. I have implemented this in the latest commit.

File: rustls-cert-gen/src/cert.rs
Comment (tbro) [reply rustls-cert-gen/src/cert.rs]: should be addressed now.

File: rustls-cert-gen/src/cert.rs
Comment (tbro) [reply rustls-cert-gen/src/cert.rs]: ditto

File: rustls-cert-gen/src/main.rs
Comment (tbro) [reply rustls-cert-gen/src/main.rs]: done

File: rustls-cert-gen/src/cert.rs
Comment (tbro) [reply rustls-cert-gen/src/cert.rs:155]: I think this is a special case. `push` is a bit misleading since `DistinquishedName` is backed by a `hashmap`. Under the hood, `entries.insert()` is called with the `DnType` as the key. So a second push will replace rather than append. So I think the behavior is ok. We can add a docstring like "Set CommonName. Multiple calls will replace previous value".

File: rustls-cert-gen/src/cert.rs
Comment (tbro) [reply rustls-cert-gen/src/cert.rs]: This change results in warnings from  `cargo doc`  causing some CI jobs to fail. The warnings are:
```
warning: public documentation for `CertificateBuilder` links to private item `EndEntity`
```
Something must be escaping my attention, but I don't quite understand the warning.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [reply rustls-cert-gen/src/cert.rs:157]: Ah! Great point. You're right :+1: Sorry about the confusion.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [reply rustls-cert-gen/src/cert.rs]: I think this is related to [feedback I left on the other PR](https://github.com/rustls/rcgen/pull/185#discussion_r1385550049).

If the intent is for this code to be usable as a library in addition to being useful through the CLI then I think in `rustls-cert-gen/src/lib.rs` where you have `pub use cert::CertificateBuilder` you should add the CA and EE types as `pub` exports.

Comment (tbro) [reply rustls-cert-gen/src/cert.rs]: Oh yes I forgot about that `lib.rs`.

**Comment by tbro** [comment]
> Thanks! Here's another round of feedback

I think I have addressed everything in that round.

**Comment by cpu** [comment]
Thank you! I will do another pass in the next day or two.

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:130]:
```suggestion
	/// Sign with `signer` and serialize.
```

File: rustls-cert-gen/src/lib.rs
Comment (cpu) [inline rustls-cert-gen/src/lib.rs:4]:
```suggestion
//! applications that verify chains of trust. It can be used for
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:152]:
```suggestion
	/// Add CommonName to `distinguished_name`. Multiple calls will
```

File: rustls-cert-gen/src/cert.rs
Comment (cpu) [inline rustls-cert-gen/src/cert.rs:74]: Re-reviewing this, I think `CaBuilder` and `EndEntityBuilder` might be better names since these are returned from the `CertificateBuilder` and continue the building process.

File: rustls-cert-gen/src/lib.rs
Comment (cpu) [inline rustls-cert-gen/src/lib.rs:8]: I think we want `EndEntityParams` and `CaParams` here too (or whatever they end up named if you agree with my rename suggestion).

Right now the return types from `CertificateBuilder::end_entity` and `CertificateBuilder::ca` are opaque out of crate.

File: rustls-cert-gen/src/main.rs
Comment (cpu) [inline rustls-cert-gen/src/main.rs:45]:
```suggestion
/// rustls-cert-gen TLS Certificate Generator
```

File: rustls-cert-gen/src/main.rs
Comment (cpu) [inline rustls-cert-gen/src/main.rs:47]:
```suggestion
	/// Output directory for generated files
```

File: rustls-cert-gen/src/main.rs
Comment (cpu) [inline rustls-cert-gen/src/main.rs:113]: I think it would be nice to have an IPv6 address testcase here too.

**Review by cpu — COMMENTED** [review]
Sorry for the delay getting this another review. I only have small nits to flag. I think there's likely good future work (like supporting a way to override the validity period), but I'm leaning towards trying to land this ASAP and iterating from there.

**Comment by tbro** [comment]
> Sorry for the delay getting this another review. I only have small nits to flag. I think there's likely good future work (like supporting a way to override the validity period), but I'm leaning towards trying to land this ASAP and iterating from there.

Yes I also thought of the validity periods. I'm sure there are a few more features that could be added. But my preference is also to finish this first bit. Then it should be much easier to review smaller future PRs that target specific features/enhancements.

**Comment by tbro** [comment]
I believe latest review comments have been addressed.

**Review by est31 — APPROVED** [review]

File: rcgen/Cargo.toml
Comment (djc) [inline rcgen/Cargo.toml:29]: Personally I would prefer specifying all dependency versions at the workspace level, because otherwise it's not very clear when adding a dependency where to look.

File: rustls-cert-gen/src/cert.rs
Comment (djc) [inline rustls-cert-gen/src/cert.rs:52]: Please add empty lines between all the methods.

File: rustls-cert-gen/src/cert.rs
Comment (djc) [inline rustls-cert-gen/src/cert.rs:32]: I think usually we link names with both backticks and brackets, like ```[`CertificateParams`]```.

File: rustls-cert-gen/src/cert.rs
Comment (djc) [inline rustls-cert-gen/src/cert.rs:59]: Should probably be ```[`Ca`]```, let's not capitalize "Certificates" here?

File: rustls-cert-gen/src/cert.rs
Comment (djc) [inline rustls-cert-gen/src/cert.rs:9]: In terms of item ordering I think it would make more sense to keep `PemCertifiedKey` closer to `EndEntity` and `Certificate`, also keeping `EndEntityBuilder` and `CaBuilder` together.

**Review by djc — COMMENTED** [review]
I have a few minor nits (hope these are not contradicting any earlier feedback), this looks good to me!

**Review by djc — APPROVED** [review]

**Comment by cpu** [comment]
> exclude rustls-cert-gen from semver check

Why is this necessary? Can you add context motivating the change to the commit message?

**Comment by tbro** [comment]
> > exclude rustls-cert-gen from semver check
> 
> Why is this necessary - because there's no initial published verison? Can you add context motivating the change to the commit message?

Yes I'm working around CI failure due to there no published version of `rustls-cert-gen`. I have updated commit message as requested.

**Comment by cpu** [comment]
Makes sense, thank you!

**Comment by tbro** [comment]
Hello everyone. Since we have 3 approvals can we go ahead and merge this? Pretty much every merge into `main` is causing conflicts requiring manual conflict resolution. In some cases changes to code. I have left these resolutions as separate commits above with clear messages, so they can easily be reviewed.  I am happy to squashed them all together given your go ahead.  I fear if this PR is left sitting around the burden will become too much to keep up with. I understand that there are a few comments remaining above, but they are self-described as "minor nits". I suggest we move those to a separate issue . Feel free to assign it to me.

**Comment by cpu** [comment]
> Hello everyone. Since we have 3 approvals can we go ahead and merge this?

That sounds good to me, I've been waiting for the branch to build green in CI again and it looks like that's been addressed. Thanks!

@djc @est31 Any reservations about merging this?

**Comment by est31** [comment]
I'm fine with merging!

**Comment by est31** [comment]
(and thanks @tbro for your contribution)

**Comment by cpu** [comment]
> I'm fine with merging!

Cool, thanks for chiming in. I'm going to put it in the queue. Talking to Djc OOB I don't think he'll have a chance to look at this. If he does later we can tackle feedback in follow-up PRs.

> (and thanks @tbro for your contribution)

+1!

**Pull request merged**
